news/_posts/2019-12-25-prestashop-in-2019-and-beyond-part-3-the-future-architecture.md

```diff
@@ -0,0 +1,269 @@
+
+In addition, a well-designed architecture should be _internally_ coherent and resilient to errors. The Core should therefore enforce coherency in its internal state and reject any possibility of incoherent input, by design. This means, among other things, strict typing and typed exceptions.
+
+#### It has to be solid, extensible and intuitive
+
+PrestaShop is a platform for building e-commerce websites, so obviously developers will need to be able to extend its behavior in order to add specific features like payment and delivery providers, ERP connectors and so on.
+
+For that, developers should have a powerful, stable, and well-documented toolset at their disposal. This includes extension points, well-defined channels for communicating with the Core, and a set of predefined modular components that encourage reusability and coherence across the ecosystem.
+
+In order to enforce systemwide coherence, the Core should be at the center of all state changes. Similarly, system state micro-management should be strongly discouraged.
+
+#### It has to be decoupled
+
+The server-side Core interface needs to become fully independent from its clients, whether it is the front-end user interface or extensions, both in the Front Office and in the Back Office. 
+
+PrestaShop is complex all on its own, but it's also a development platform. In order to be flexible, it has to be made out a large number of small, independent components, instead of the other way around. 
+
+#### It cannot be everything at once
+
+I believe that software quality can be measured in 4 key dimensions:
+
+- **Simplicity / Ease of use:** How easy it is to work with it
+- **Flexibility:** How customizable it is, how well can it adapt or be adapted to your needs
+- **Reliability:** How stable it is, how prone it is to behave unexpectedly or perform below expectations
+- **Features:** How powerful it is, how many things it can do
+
+An ideal software should be good in all four. But let's be honest: in the real world, it's not possible to be excel in everything at once. In order really perform at a high level, some trade-offs need to be made, and it usually means focusing on a smaller number of things.
+
+Much the same as the ["Fast-Good-Cheap, pick two" problem](https://fastgood.cheap/), in my opinion, software can only be realistically optimized for three of these quality axes at a time, and one must always be sacrificed in the interest of the others. Here's a graphical representation:
+
+![Choose your trade-off][architecture-tradeoffs]
+
+If we are forced to choose between one of these archetypes, which one would we be rather be? Not buggy, that's for sure. Not proprietary either, because PrestaShop is an open source platform. We have to choose between having very few features but do them really well, or have many features at the cost of a steeper learning curve.
```

Comment:
I am not convinced by this demonstration and I don't see why you need it here ?

Reply:
I've heard people complain that PrestaShop 1.7 is too complex compared with 1.6. They need to understand what they're getting in exchange for that perceived complexity.